api/provenance/provenance.go

```diff
@@ -6,57 +6,73 @@ package provenance
 import (
 	"fmt"
 	"runtime"
+	"runtime/debug"
 	"strings"
 )
 
+// These variables are set at build time using ldflags.
+//
+//nolint:gochecknoglobals
 var (
-	version = "unknown"
-	// sha1 from git, output of $(git rev-parse HEAD)
-	gitCommit = "$Format:%H$"
+	// During a release, this will be set to the release tag, e.g. "kustomize/v4.5.7"
+	version = developmentVersion
 	// build date in ISO8601 format, output of $(date -u +'%Y-%m-%dT%H:%M:%SZ')
-	buildDate = "1970-01-01T00:00:00Z"
-	goos      = runtime.GOOS
-	goarch    = runtime.GOARCH
+	buildDate = "unknown"
 )
 
+// This default value, (devel), matches
+// the value debug.BuildInfo uses for an unset main module version.
+const developmentVersion = "(devel)"
+
 // Provenance holds information about the build of an executable.
 type Provenance struct {
 	// Version of the kustomize binary.
-	Version string `json:"version,omitempty"`
+	Version string `json:"version,omitempty" yaml:"version,omitempty"`
 	// GitCommit is a git commit
-	GitCommit string `json:"gitCommit,omitempty"`
+	GitCommit string `json:"gitCommit,omitempty" yaml:"gitCommit,omitempty"`
 	// BuildDate is date of the build.
-	BuildDate string `json:"buildDate,omitempty"`
+	BuildDate string `json:"buildDate,omitempty" yaml:"buildDate,omitempty"`
 	// GoOs holds OS name.
-	GoOs string `json:"goOs,omitempty"`
+	GoOs string `json:"goOs,omitempty" yaml:"goOs,omitempty"`
 	// GoArch holds architecture name.
-	GoArch string `json:"goArch,omitempty"`
+	GoArch string `json:"goArch,omitempty" yaml:"goArch,omitempty"`
+	// GoVersion holds Go version.
+	GoVersion string `json:"goVersion,omitempty" yaml:"goVersion,omitempty"`
 }
 
 // GetProvenance returns an instance of Provenance.
 func GetProvenance() Provenance {
-	return Provenance{
-		version,
-		gitCommit,
-		buildDate,
-		goos,
-		goarch,
+	p := Provenance{
+		BuildDate: buildDate,
+		Version:   version,
+		GitCommit: "unknown",
+		GoOs:      runtime.GOOS,
+		GoArch:    runtime.GOARCH,
+		GoVersion: runtime.Version(),
+	}
+	info, ok := debug.ReadBuildInfo()
+	if !ok {
+		return p
+	}
+
+	for _, setting := range info.Settings {
+		// For now, the git commit is the only information of interest.
+		// We could consider adding other info such as the commit date in the future.
+		if setting.Key == "vcs.revision" {
+			p.GitCommit = setting.Value
+		}
 	}
+	return p
 }
 
 // Full returns the full provenance stamp.
 func (v Provenance) Full() string {
 	return fmt.Sprintf("%+v", v)
 }
 
-// Short returns the shortened provenance stamp.
+// Short returns the semantic version.
 func (v Provenance) Short() string {
-	return fmt.Sprintf(
-		"%v",
-		Provenance{
-			Version:   v.Version,
-			BuildDate: v.BuildDate,
-		})
+	return v.Semver()
```

Comment:
technically, this is already breaking. Alternatively, we could either leave it as is (and update the deprecation notice on the main format), or go ahead and remove it in v5 without deprecation. Wdyt @natasha41575 ?

Reply:
Will leave this to @natasha41575